// Device service APIs

syntax = "proto3";
package distributed_match_engine;

import "loc.proto";

message Match_Engine_Request {
  uint32 ver = 1;
  // User ID type - IMEI, MSISDN etc
  enum IDType {
	  IMEI = 0;
	  MSISDN = 1;
	  IPADDR = 2;
  }
  IDType IdType = 2;
  // Actual ID
  string Id = 3;
  // The carrier that user is connected to
  uint64 Carrier = 4;
  // The tower that user is currently connected to
  uint64 Tower = 5;
  // The GPS location of the user 
  edgeproto.Loc GpsLocation = 6;
  // Edge-cloud assigned application ID
  uint64 AppId = 7;
  // Protocol application uses
  bytes Protocol = 8;
  // The protocol port on the server side
  bytes ServerPort = 9;
  // App Developer Name
  string DevName = 10;
  // App Name
  string AppName = 11;
  // App Version
  string AppVers = 12;
}

message Match_Engine_Reply {
  uint32 ver = 1;
  // domain name
  string uri = 2;
  // ip of the app service
  bytes service_ip = 3;
  // port of the app service?
  uint32 service_port = 4;
  // location of the cloudlet?
  edgeproto.Loc cloudlet_location = 5;
  // Status return
  bool Status = 6;
}

message Match_Engine_Loc_Verify {
  uint32 ver = 1;
  // Status of the reply
  enum Tower_Status {
    UNKNOWN = 0;
    CONNECTED_TO_SPECIFIED_TOWER = 1;
    NOT_CONNECTED_TO_SPECIFIED_TOWER = 2;
  }
  Tower_Status tower_status = 2;
  enum GPS_Location_Status {
    LOC_UNKNOWN = 0;
    LOC_WITHIN_1M = 1;
    LOC_WITHIN_10M = 2;
  }
  GPS_Location_Status gps_location_status = 3;
}

service Match_Engine_Api {
  rpc FindCloudlet(Match_Engine_Request) returns (Match_Engine_Reply) {}
  rpc VerifyLocation(Match_Engine_Request) returns (Match_Engine_Loc_Verify) {}
}


